File: volume_1/chapter_7/llamaindex/llm_tools_gateway.js
/**
 * LLM Tools - LlamaIndex JS with Wikipedia.
 */

import { LlamaIndexLLMManager as Chapter6LlamaIndexManager } from '../../chapter_6/llamaindex/llm_memory_retrieval.js';
import { interactiveCli, getDefaultModel } from '../../chapter_4/utils.js';
import { normalizeResponseText } from '../../chapter_4/stream.js';
import { buildToolsPrompt, runTool } from '../tools.js';

const TOOLS_TEMPLATE = `You are a helpful assistant with access to external tools.

Available tools:
{tools}

Return strict JSON:
{
  "tool_call": null OR {"name": "tool_name", "arguments": {"arg": "value"}},
  "final_answer": "string"
}

Rules:
- If no tool is needed, set tool_call to null.
- If a tool is needed, set tool_call and keep final_answer short.
- Return JSON only.

User topic: {topic}`;

const FOLLOW_UP_TEMPLATE = `You already requested a tool and now have the result.

Original user topic: {topic}
Tool call: {tool_call}
Tool output: {tool_output}

Return strict JSON:
{
  "tool_call": null,
  "final_answer": "final response for the user"
}`;

class LlamaIndexLLMManager extends Chapter6LlamaIndexManager {
    constructor(memoryEnabled = true, retrievalK = 4) {
        super(memoryEnabled, retrievalK);
        this.framework = 'LlamaIndex Tools JS';
    }

    _extractJsonObject(raw) {
        let text = String(raw || '').trim();
        if (text.startsWith('```json')) text = text.slice(7);
        if (text.startsWith('```')) text = text.slice(3);
        if (text.endsWith('```')) text = text.slice(0, -3);
        text = text.trim();

        try {
            return JSON.parse(text);
        } catch {
            const match = text.match(/\{[\s\S]*\}/);
            if (!match) throw new Error('No JSON object found in model response');
            return JSON.parse(match[0]);
        }
    }


    _buildFallbackToolPayload(rawText) {
        const fallbackAnswer = normalizeResponseText(rawText).trim();
        if (!fallbackAnswer) {
            throw new Error('No JSON object found in model response');
        }
        return {
            tool_call: null,
            final_answer: fallbackAnswer,
        };
    }

    _normalizeToolPayload(rawPayload, rawText) {
        if (!rawPayload || typeof rawPayload !== 'object' || Array.isArray(rawPayload)) {
            return this._buildFallbackToolPayload(rawText);
        }

        const toolCall = rawPayload.tool_call;
        const finalAnswer = rawPayload.final_answer;

        if (typeof finalAnswer === 'string' && finalAnswer.trim()) {
            return {
                tool_call: toolCall && typeof toolCall === 'object' ? toolCall : null,
                final_answer: finalAnswer.trim(),
            };
        }

        return this._buildFallbackToolPayload(rawText);
    }

    async _invokeJsonStep(provider, prompt, temperature, maxTokens) {
        const client = this._createClient(provider, temperature, maxTokens);
        const result = await client.chat({ messages: [{ role: 'user', content: prompt }] });
        const text = this._extractText(result);

        try {
            const payload = this._extractJsonObject(text);
            return { payload: this._normalizeToolPayload(payload, text), result };
        } catch {
            return { payload: this._buildFallbackToolPayload(text), result };
        }
    }

    async _buildRetrievalContext(provider, topic, sessionId) {
        const messages = this.retrievalMemoryEnabled
            ? await this._getMemoryMessages(provider, sessionId)
            : [];
        const retrieved = this.retrievalMemoryEnabled
            ? this._selectRetrievedMessages(topic, messages)
            : [];

        const retrievedContext = retrieved.map((item) => `[${item.role}] ${item.content}`).join('\n');
        const retrievalAugmentedTopic = retrievedContext
            ? `Relevant memory snippets:\n${retrievedContext}\n\nCurrent user topic: ${topic}`
            : topic;

        const fullHistoryContext = messages
            .map((msg) => `[${String(msg?.role ?? 'unknown')}] ${String(msg?.content ?? '')}`)
            .join('\n');
        const promptWithoutRetrieval = fullHistoryContext
            ? `${fullHistoryContext}\n\nCurrent user topic: ${topic}`
            : topic;

        const tokensWithRetrieval = this._estimateTokens(retrievalAugmentedTopic);
        const tokensWithoutRetrieval = this._estimateTokens(promptWithoutRetrieval);
        const estimatedSaved = Math.max(0, tokensWithoutRetrieval - tokensWithRetrieval);
        const reductionPercent = tokensWithoutRetrieval > 0
            ? Number(((estimatedSaved / tokensWithoutRetrieval) * 100).toFixed(2))
            : 0;

        return {
            retrievalAugmentedTopic,
            retrievalMetadata: {
                history_messages_available: messages.length,
                retrieved_messages_count: retrieved.length,
                retrieved_messages: retrieved,
                tokens_with_memory_retrieval: tokensWithRetrieval,
                tokens_without_memory_retrieval: tokensWithoutRetrieval,
                estimated_tokens_saved: estimatedSaved,
                estimated_token_reduction_percent: reductionPercent,
            },
        };
    }



    _resolveToolsTemplate(template) {
        const candidate = String(template ?? '').trim();
        if (!candidate || candidate === '{topic}' || !candidate.includes('{tools}')) {
            return TOOLS_TEMPLATE;
        }
        return template;
    }

    _normalizeWikipediaQuery(topic) {
        const text = String(topic || '').trim();
        if (!text) return text;
        return text
            .replace(/^\s*(what\s+is|who\s+is|where\s+is|when\s+did|when\s+was|why\s+is|how\s+is)\s+/i, '')
            .replace(/[?]+$/g, '')
            .trim();
    }

    _shouldForceWikipediaTool(topic, toolCall) {
        if (toolCall && typeof toolCall === 'object' && toolCall.name) return false;
        const text = String(topic || '').toLowerCase();
        if (!text.trim()) return false;

        const creativeSignals = [
            'poem', 'story', 'fiction', 'brainstorm', 'imagine', 'creative writing', 'roleplay', 'joke',
        ];
        if (creativeSignals.some((k) => text.includes(k))) return false;

        const factualSignals = [
            'what is', 'who is', 'when did', 'where is', 'why', 'how', 'define', 'history', 'date', 'science',
        ];
        return factualSignals.some((k) => text.includes(k)) || text.split(/\s+/).length >= 3;
    }

    async askQuestion(topic, provider = null, template = TOOLS_TEMPLATE, maxTokens = 1000, temperature = 0.2, sessionId = 'default') {
        template = this._resolveToolsTemplate(template);
        const resolvedProvider = this._resolveProvider(provider);
        const basePrompt = template.replace('{topic}', topic).replace('{tools}', buildToolsPrompt());

        if (!resolvedProvider) {
            return {
                success: false,
                error: 'No providers available',
                provider: 'none',
                model: 'none',
                prompt: basePrompt,
                response: null,
            };
        }

        const { retrievalAugmentedTopic, retrievalMetadata } = await this._buildRetrievalContext(resolvedProvider, topic, sessionId);
        const prompt = template.replace('{topic}', retrievalAugmentedTopic).replace('{tools}', buildToolsPrompt());
        const model = getDefaultModel(resolvedProvider);

        try {
            const { payload: firstStep } = await this._invokeJsonStep(resolvedProvider, prompt, temperature, maxTokens);
            const toolCall = firstStep.tool_call;
            let finalAnswer = String(firstStep.final_answer || '').trim();
            let toolOutput = null;

            let effectiveToolCall = toolCall;
            if (this._shouldForceWikipediaTool(topic, effectiveToolCall)) {
                effectiveToolCall = {
                    name: 'get_wikipedia_evidence_pack',
                    arguments: { query: this._normalizeWikipediaQuery(topic) || topic },
                };
            }

            if (effectiveToolCall && typeof effectiveToolCall === 'object' && effectiveToolCall.name) {
                const toolName = String(effectiveToolCall.name);
                const toolArgs = effectiveToolCall.arguments && typeof effectiveToolCall.arguments === 'object' ? effectiveToolCall.arguments : {};
                toolOutput = await runTool(toolName, toolArgs);

                const followUpPrompt = FOLLOW_UP_TEMPLATE
                    .replace('{topic}', topic)
                    .replace('{tool_call}', JSON.stringify(effectiveToolCall))
                    .replace('{tool_output}', String(toolOutput));

                const { payload: secondStep } = await this._invokeJsonStep(resolvedProvider, followUpPrompt, temperature, maxTokens);
                finalAnswer = String(secondStep.final_answer || finalAnswer).trim() || finalAnswer;
            }

            const rawResponse = JSON.stringify({
                tool_call: effectiveToolCall ?? null,
                tool_output: toolOutput,
                final_answer: finalAnswer,
            });

            const responsePayload = {
                tool_call: effectiveToolCall ?? null,
                tool_output: toolOutput,
                final_answer: finalAnswer,
                metadata: {
                    ...this._buildMetadata({
                        provider: resolvedProvider,
                        model,
                        sessionId,
                        temperature,
                        maxTokens,
                    }, rawResponse),
                    retrieval: retrievalMetadata,
                },
            };

            if (this.retrievalMemoryEnabled) {
                await this._appendToMemory(resolvedProvider, sessionId, 'user', topic);
                await this._appendToMemory(resolvedProvider, sessionId, 'assistant', rawResponse);
                await this._persistMemory(resolvedProvider, sessionId);
            }

            return {
                success: true,
                provider: resolvedProvider,
                model,
                prompt,
                response: responsePayload,
                rawAnswer: finalAnswer,
                temperature,
                maxTokens,
                sessionId,
            };
        } catch (error) {
            return {
                success: false,
                provider: resolvedProvider,
                model,
                prompt,
                error: error.message,
                response: null,
                temperature,
                maxTokens,
                sessionId,
            };
        }
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.length > 0 && args[0] === 'web') {
        const { runWebServer } = await import('../../chapter_4/web.js');
        await runWebServer(() => new LlamaIndexLLMManager(true));
    } else {
        const manager = new LlamaIndexLLMManager(true);
        await manager._checkProviders();
        await interactiveCli(manager);
    }
}

export { LlamaIndexLLMManager };

if (import.meta.url === `file://${process.argv[1]}`) {
    main().catch(console.error);
}

File: volume_1/chapter_7/llamaindex/llm_tools_gateway.py
"""LLM Tools Gateway - LlamaIndex (Chapter 7).

Builds directly on Chapter 6 retrieval-memory manager hierarchy:
Chapter 4 (base providers) -> Chapter 5 (memory/persistence) ->
Chapter 6 (retrieval memory + structured output) -> Chapter 7 (tool use).
"""

from __future__ import annotations

import json
import os
import re
import sys
from typing import Dict, Optional, Tuple

from llama_index.core.llms import ChatMessage

CHAPTER_4_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "chapter_4"))
CHAPTER_6_LLAMAINDEX = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "chapter_6", "llamaindex"))
CHAPTER_7_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))

sys.path.append(CHAPTER_4_ROOT)
sys.path.append(CHAPTER_6_LLAMAINDEX)
sys.path.append(CHAPTER_7_ROOT)

from llm_memory_retrieval_gateway import LlamaIndexLLMManager as Chapter6LlamaIndexManager
from tools import build_tools_prompt, run_tool
from utils import get_default_model, interactive_cli

TOOLS_TEMPLATE = """
You are a helpful assistant with access to external tools.

Available tools:
{tools}

Return strict JSON:
{{
  "tool_call": null OR {{"name": "tool_name", "arguments": {{"arg": "value"}}}},
  "final_answer": "string"
}}

Rules:
- If no tool is needed, set tool_call to null.
- If a tool is needed, set tool_call and keep final_answer short.
- Return JSON only.

User topic: {topic}
""".strip()

FOLLOW_UP_TEMPLATE = """
You already requested a tool and now have the result.

Original user topic: {topic}
Tool call: {tool_call}
Tool output: {tool_output}

Return strict JSON:
{{
  "tool_call": null,
  "final_answer": "final response for the user"
}}
""".strip()


class LlamaIndexLLMManager(Chapter6LlamaIndexManager):
    """Chapter 7 LlamaIndex manager with retrieval-aware tool orchestration."""

    def __init__(self, memory_enabled: bool = True, retrieval_k: int = 4):
        super().__init__(memory_enabled=memory_enabled, retrieval_k=retrieval_k)
        self.framework = "LlamaIndex+Memory+Retrieval+Tools"

    @staticmethod
    def _extract_json_object(raw: str) -> Dict:
        text = raw.strip()
        if text.startswith("```json"):
            text = text[7:]
        if text.startswith("```"):
            text = text[3:]
        if text.endswith("```"):
            text = text[:-3]

        text = text.strip()
        try:
            return json.loads(text)
        except json.JSONDecodeError:
            match = re.search(r"\{[\s\S]*\}", text)
            if not match:
                raise
            return json.loads(match.group(0))

    def _invoke_json_step(self, provider: str, prompt: str, temperature: float, max_tokens: int) -> Tuple[Dict, object]:
        client = self._create_client(provider, temperature=temperature, max_tokens=max_tokens)
        result = client.chat([ChatMessage(role="user", content=prompt)])
        text = self._extract_text(result)
        return self._extract_json_object(text), result

    def _build_retrieval_context(self, provider: str, topic: str, session_id: str) -> Tuple[str, Dict[str, object]]:
        memory = self._get_memory(provider, session_id) if self.retrieval_memory_enabled else None
        messages = self._memory_messages(memory) if memory else []
        retrieved = self._select_retrieved_messages(topic, messages) if self.retrieval_memory_enabled else []

        retrieved_context = "\n".join(f"[{item['role']}] {item['content']}" for item in retrieved)
        retrieval_augmented_topic = (
            f"Relevant memory snippets:\n{retrieved_context}\n\nCurrent user topic: {topic}" if retrieved_context else topic
        )

        full_history_context = "\n".join(
            f"[{getattr(msg, 'role', 'unknown')}] {getattr(msg, 'content', '')}" for msg in messages
        )
        prompt_without_retrieval = (
            f"{full_history_context}\n\nCurrent user topic: {topic}" if full_history_context else topic
        )

        tokens_with_retrieval = self._estimate_tokens(retrieval_augmented_topic)
        tokens_without_retrieval = self._estimate_tokens(prompt_without_retrieval)
        estimated_saved = max(0, tokens_without_retrieval - tokens_with_retrieval)
        reduction_percent = round((estimated_saved / tokens_without_retrieval) * 100, 2) if tokens_without_retrieval else 0.0

        retrieval_metadata = {
            "history_messages_available": len(messages),
            "retrieved_messages_count": len(retrieved),
            "retrieved_messages": retrieved,
            "tokens_with_memory_retrieval": tokens_with_retrieval,
            "tokens_without_memory_retrieval": tokens_without_retrieval,
            "estimated_tokens_saved": estimated_saved,
            "estimated_token_reduction_percent": reduction_percent,
        }
        return retrieval_augmented_topic, retrieval_metadata

    @staticmethod
    def _resolve_tools_template(template: Optional[str]) -> str:
        candidate = (template or "").strip()
        if not candidate or candidate == "{topic}" or "{tools}" not in candidate:
            return TOOLS_TEMPLATE
        return template

    def ask_question(
        self,
        topic: str,
        provider: Optional[str] = None,
        template: str = TOOLS_TEMPLATE,
        max_tokens: int = 1000,
        temperature: float = 0.2,
        session_id: str = "default",
    ) -> Dict:
        template = self._resolve_tools_template(template)
        provider = self._resolve_provider(provider)
        prompt = template.format(topic=topic, tools=build_tools_prompt())

        if not provider:
            return {
                "success": False,
                "error": "No providers available",
                "provider": "none",
                "model": "none",
                "prompt": prompt,
                "response": None,
            }

        model = get_default_model(provider)
        retrieval_topic, retrieval_metadata = self._build_retrieval_context(provider, topic, session_id)
        retrieval_prompt = template.format(topic=retrieval_topic, tools=build_tools_prompt())

        try:
            first_step, _ = self._invoke_json_step(provider, retrieval_prompt, temperature=temperature, max_tokens=max_tokens)
            tool_call = first_step.get("tool_call")
            final_answer = str(first_step.get("final_answer", "")).strip()

            tool_output = None
            if isinstance(tool_call, dict) and tool_call.get("name"):
                tool_name = str(tool_call.get("name"))
                tool_args = tool_call.get("arguments") or {}
                if not isinstance(tool_args, dict):
                    tool_args = {}

                tool_output = run_tool(tool_name, tool_args)
                follow_up_prompt = FOLLOW_UP_TEMPLATE.format(
                    topic=topic,
                    tool_call=json.dumps(tool_call, ensure_ascii=False),
                    tool_output=tool_output,
                )
                second_step, _ = self._invoke_json_step(provider, follow_up_prompt, temperature=temperature, max_tokens=max_tokens)
                final_answer = str(second_step.get("final_answer", final_answer)).strip() or final_answer

            raw_response = json.dumps(
                {
                    "tool_call": tool_call,
                    "tool_output": tool_output,
                    "final_answer": final_answer,
                },
                ensure_ascii=False,
            )

            payload = {
                "tool_call": tool_call,
                "tool_output": tool_output,
                "final_answer": final_answer,
                "metadata": {
                    **self._build_metadata(
                        {
                            "provider": provider,
                            "model": model,
                            "session_id": session_id,
                            "temperature": temperature,
                            "max_tokens": max_tokens,
                        },
                        raw_response,
                    ),
                    "retrieval": retrieval_metadata,
                },
            }

            if self.retrieval_memory_enabled:
                self._append_to_memory(provider, session_id, "user", topic)
                self._append_to_memory(provider, session_id, "assistant", raw_response)
                self._persist_memory(provider, session_id)

            return {
                "success": True,
                "provider": provider,
                "model": model,
                "prompt": retrieval_prompt,
                "response": payload,
                "raw_answer": final_answer,
                "temperature": temperature,
                "max_tokens": max_tokens,
                "session_id": session_id,
            }
        except Exception as exc:
            return {
                "success": False,
                "provider": provider,
                "model": model,
                "prompt": retrieval_prompt,
                "error": str(exc),
                "response": None,
                "temperature": temperature,
                "max_tokens": max_tokens,
                "session_id": session_id,
            }


def main():
    if len(sys.argv) > 1 and sys.argv[1] == "web":
        from web import run_web_server

        run_web_server(lambda: LlamaIndexLLMManager(memory_enabled=True))
    else:
        interactive_cli(LlamaIndexLLMManager(memory_enabled=True))


if __name__ == "__main__":
    main()
